app/src/main/java/io/appium/settings/LocationService.java

```diff
@@ -40,28 +77,160 @@ public int onStartCommand(Intent intent, int flags, int startId) {
                 return START_NOT_STICKY;
             }
         }
+        handleIntent(intent);
+
+        return START_NOT_STICKY;
+    }
+
+    @Override
+    public void onDestroy() {
+        Log.i(TAG, "Shutting down MockLocationService");
+        locationUpdatesTimer.cancel();
+        disableLocationProviders();
+        super.onDestroy();
+    }
+
+    private void handleIntent(Intent intent) {
+        if (intent == null) {
+            return;
+        }
+        // update the locationFactory also if the service is already running to mock
+        updateMockLocationFactory(intent);
+        scheduleLocationUpdate();
+    }
+
+    private void enableLocationProviders() {
+        for (MockLocationProvider mockLocationProvider : mockLocationProviders) {
+            try {
+                mockLocationProvider.enable();
+            } catch (Exception e) {
+                Log.e(TAG, String.format("Couldn't enable location provider: '%s'",
+                        mockLocationProvider.getProviderName()));
+            }
+        }
+    }
+
+    private void disableLocationProviders() {
+        for (MockLocationProvider mockLocationProvider : mockLocationProviders) {
+            try {
+                mockLocationProvider.disable();
+            } catch (Exception e) {
+                Log.e(TAG, String.format("Could not disable location provider: '%s'",
+                        mockLocationProvider.getProviderName()));
+            }
+        }
+    }
+
+    private void initializeLocationProviders() {
+        LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
+
+        mockLocationProviders.addAll(createMockProviders(locationManager));
```

Comment:
is it guaranteed this method is going to be called only once?
IMHO it would be safer to perform clear on `mockLocationProviders`